src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala

```diff
@@ -93,6 +96,581 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
+        // TX2 modified table all of which was read by TX1
+        tx1.commit(addC_2_1_nested :: Nil, Truncate())
+      }
+    }
+  }
+
+  test("block commit when part. range read conflicts with concur. write in lvl2 nested partition") {
+    withLog(
+      addA_1_1_nested :: Nil,
+      partitionCols = "a" :: "b" :: Nil) { log =>
+
+      val tx1 = log.startTransaction()
+      // TX1 reads multiple nested partitions a >= 1 or b > 1
+      val tx1Read = tx1.filterFiles(('a >= 1 or 'b > 1).expr :: Nil)
+      assert(tx1Read.map(_.path).toSet == Set(A_1_1))
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies a=1/b=2
+      tx2.commit(addB_1_2_nested :: Nil, Truncate())
+
+      intercept[ConcurrentModificationException] {
+        // partition a=1/b=2 conflicts with our read a >= 1 or 'b > 1
+        tx1.commit(addD_3_1_nested :: Nil, Truncate())
+      }
+    }
+  }
+
+  test("block commit with concurrent removes on same file") {
+    withLog(addB_1_2_nested :: Nil, partitionCols = "a" :: "b" :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 reads a=2 so that read is disjoint with write partition.
+      tx1.filterFiles(('a === 2).expr :: Nil)
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies a=1/b=2
+      tx2.commit(addB_1_2_nested.remove :: Nil, Truncate())
+
+      intercept[ConcurrentModificationException] {
+        // TX1 read does not conflict with TX2 as disjoint partitions
+        // But TX2 removed the same file that TX1 is trying to remove
+        tx1.commit(addB_1_2_nested.remove:: Nil, Truncate())
+      }
+    }
+  }
+
+  /*
+  test("block concurrent commit on Delete attempt and Add initially empty") {
+    withConcurrentLog(addC_P2 :: addE_P3 :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // empty read
+      val tx1Read = tx1.filterFiles(('part === 1).expr :: Nil)
+      assert(tx1Read.isEmpty)
+      // tx2 deletes "part=p1"
+      val tx2 = log.startTransaction()
+      // empty read
+      val tx2Read = tx2.filterFiles(('part === 1).expr :: Nil)
+      assert(tx2Read.isEmpty)
+
+      tx1.commit(addA_P1 :: Nil, Truncate())
+
+      intercept[ConcurrentModificationException] {
+        // tx1 modified P1 which we tried to read and could try to delete
+        tx2.commit(Seq.empty, Truncate())
+      }
+    }
+  }
+  */
+
+  test("allow concurrent commit on Delete and Add disjoint partitions initially empty") {
+    withLog(addC_P2 :: addE_P3 :: Nil) { log =>
+      // tx1 deletes "part=p1"
+      val tx1 = log.startTransaction()
+      // empty read
+      val tx1Read = tx1.filterFiles(('part === 1).expr :: Nil)
+      assert(tx1Read.isEmpty)
+
+      // tx2 deletes "part=p1"
+      val tx2 = log.startTransaction()
+      // empty read
+      val tx2Read = tx2.filterFiles(('part === 1).expr :: Nil)
+      assert(tx2Read.isEmpty)
+
+      tx1.commit(addD_P2 :: Nil, Truncate())
```

Comment:
Should be `addD_P2.remove` since `truncate` should just remove files.

Reply:
looking at the existing tests, i think Truncate() is just a place holder operation. could have been anything. 

we should probably change all of it to use ManualUpdate

Reply:
Correct, the use of Truncate() was just took over from existing tests.